src/batch/src/executor2/top_n.rs

```diff
@@ -131,44 +129,40 @@ impl TopNExecutor {
     }
 }
 
-#[async_trait::async_trait]
-impl Executor for TopNExecutor {
-    async fn open(&mut self) -> Result<()> {
-        self.child.open().await?;
-
-        while let Some(chunk) = self.child.next().await? {
-            self.top_n_heap.fit(Arc::new(chunk));
-        }
-        self.child.close().await?;
-
-        Ok(())
+impl Executor2 for TopNExecutor2 {
+    fn schema(&self) -> &Schema {
+        self.child.schema()
     }
 
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
-        if let Some(chunk) = self.top_n_heap.dump() {
-            Ok(Some(chunk))
-        } else {
-            Ok(None)
-        }
+    fn identity(&self) -> &str {
+        &self.identity
     }
 
-    async fn close(&mut self) -> Result<()> {
-        Ok(())
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
     }
+}
 
-    fn schema(&self) -> &Schema {
-        self.child.schema()
-    }
+impl TopNExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        #[for_await]
+        for data_chunk in self.child.execute() {
+            let data_chunk = data_chunk?;
+            self.top_n_heap.fit(Arc::new(data_chunk));
+        }
 
-    fn identity(&self) -> &str {
-        &self.identity
+        if let Some(chunk) = self.top_n_heap.dump() {
+            yield chunk
```

Comment:
We should limit chunk size of each executor, and this is the motivation to use `futures-async-stream` to refactor executor.